File: cclms/call_centre_lead_management_system/page/signs_dashboard/signs_dashboard.js
frappe.pages["signs-dashboard"].on_page_load = function (wrapper) {
  const page = frappe.ui.make_app_page({
    parent: wrapper,
    title: __("Signs & Attribution Dashboard"),
    single_column: true,
  });
  new SignsDashboard(page);
};

class SignsDashboard {
  constructor(page) {
    this.page = page;
    this._makeFilters();
    this._makeLayout();
    this.refresh();
  }

  // ── Filters ────────────────────────────────────────────────────────────
  _makeFilters() {
    const today = frappe.datetime.get_today();
    const mStart = today.slice(0, 7) + "-01";
    const mEnd   = today;

    this.f_from = this.page.add_field({
      label: __("From Date"), fieldtype: "Date", fieldname: "from_date", default: mStart,
      change: () => this.refresh(),
    });
    this.f_to = this.page.add_field({
      label: __("To Date"), fieldtype: "Date", fieldname: "to_date", default: mEnd,
      change: () => this.refresh(),
    });
    this.f_company = this.page.add_field({
      label: __("Company"), fieldtype: "Link", fieldname: "company",
      options: "Operator Companies", change: () => this.refresh(),
    });
    this.f_branch = this.page.add_field({
      label: __("Branch"), fieldtype: "Link", fieldname: "branch",
      options: "Branch", change: () => this.refresh(),
    });
    this.page.set_primary_action(__("Refresh"), () => this.refresh());
  }

  _args() {
    return {
      from_date: this.f_from.get_value() || null,
      to_date:   this.f_to.get_value()   || null,
      company:   this.f_company.get_value() || null,
      branch:    this.f_branch.get_value()  || null,
    };
  }

  // ── Layout skeleton ────────────────────────────────────────────────────
  _makeLayout() {
    this.page.main.html(`
      <div class="sd-root" style="padding:0 4px;">

        <!-- Summary cards -->
        <div class="sd-cards" style="display:grid;grid-template-columns:repeat(auto-fill,minmax(160px,1fr));gap:12px;margin-bottom:16px;"></div>

        <!-- Row 2: velocity bar + pipeline snapshot -->
        <div style="display:grid;grid-template-columns:1fr 1fr;gap:12px;margin-bottom:16px;">
          <div class="sd-panel">
            <div class="sd-panel-title">${__("Avg Days per Stage (from Ledger)")}</div>
            <div class="sd-velocity"></div>
          </div>
          <div class="sd-panel">
            <div class="sd-panel-title">${__("Live Pipeline Snapshot")}</div>
            <div class="sd-pipeline"></div>
          </div>
        </div>

        <!-- Row 3: approval/rejection trend -->
        <div class="sd-panel" style="margin-bottom:16px;">
          <div class="sd-panel-title">${__("Monthly Approval / Rejection / Signed Trend")}</div>
          <div class="sd-trend" style="height:260px;"></div>
        </div>

        <!-- Row 4: attribution table + pending signs -->
        <div style="display:grid;grid-template-columns:1fr 1fr;gap:12px;margin-bottom:16px;">
          <div class="sd-panel">
            <div class="sd-panel-title">${__("Agent Attribution")}</div>
            <div class="sd-attribution"></div>
          </div>
          <div class="sd-panel">
            <div class="sd-panel-title">${__("Signs Pending Commission Attribution")}
              <span class="sd-pending-badge" style="background:#ef4444;color:#fff;border-radius:999px;padding:1px 8px;font-size:11px;margin-left:6px;">0</span>
            </div>
            <div class="sd-pending"></div>
          </div>
        </div>

      </div>
    `);

    // Style helper
    this.page.main.find(".sd-panel").css({
      border: "1px solid var(--border-color)",
      borderRadius: "10px",
      padding: "12px",
      background: "#fff",
    });
    this.page.main.find(".sd-panel-title").css({
      fontWeight: "600",
      fontSize: "13px",
      marginBottom: "10px",
      color: "var(--text-color)",
    });
  }

  // ── Refresh all sections ───────────────────────────────────────────────
  async refresh() {
    const args = this._args();
    const base = "cclms.call_centre_lead_management_system.page.signs_dashboard.signs_dashboard";

    this._showLoading();

    try {
      const [summary, velocity, pipeline, trend, attribution] = await Promise.all([
        frappe.call({ method: `${base}.get_signs_summary`,            args }).then(r => r.message || {}),
        frappe.call({ method: `${base}.get_stage_velocity`,           args }).then(r => r.message || []),
        frappe.call({ method: `${base}.get_pipeline_snapshot`,        args }).then(r => r.message || {}),
        frappe.call({ method: `${base}.get_approval_rejection_trend`, args: { months_back: 6, company: args.company } }).then(r => r.message || {}),
        frappe.call({ method: `${base}.get_agent_attribution`,        args }).then(r => r.message || []),
      ]);

      this._renderCards(summary, pipeline);
      this._renderVelocity(velocity);
      this._renderPipeline(pipeline);
      this._renderTrend(trend);
      this._renderAttribution(attribution);
      this._renderPending(summary.pending_records || []);
    } catch (e) {
      console.error("Signs dashboard error", e);
      frappe.msgprint(__("Failed to load dashboard data. Check console."), __("Error"));
    }
  }

  _showLoading() {
    this.page.main.find(".sd-cards").html(
      `<div style="grid-column:1/-1;color:var(--text-muted);padding:16px">${frappe.utils.icon("loading", "sm")} ${__("Loading…")}</div>`
    );
  }

  // ── Cards ──────────────────────────────────────────────────────────────
  _renderCards(summary, pipeline) {
    const COLORS = {
      "Signs This Period": "#10b981",
      "Pending Attribution": "#ef4444",
      "Attribution Done": "#22c55e",
      "Sign Rate": "#f59e0b",
      "Rejection Rate": "#ef4444",
      "Total Active Leads": "#3b82f6",
    };
    const items = [
      { label: __("Signs This Period"),   value: summary.total_signs || 0 },
      { label: __("Pending Attribution"), value: summary.pending_attribution || 0, sub: __("no closing agent") },
      { label: __("Attribution Done"),    value: summary.attributed || 0 },
      { label: __("Sign Rate"),           value: (pipeline.sign_rate || 0) + "%" },
      { label: __("Rejection Rate"),      value: (pipeline.rejection_rate || 0) + "%" },
      { label: __("Total Active Leads"),  value: pipeline.total || 0 },
    ];

    const $c = this.page.main.find(".sd-cards").empty();
    items.forEach(({ label, value, sub }) => {
      const color = COLORS[label] || "#334155";
      $c.append(`
        <div style="border:1px solid var(--border-color);border-radius:10px;padding:14px;background:#fff;">
          <div style="font-size:11px;color:var(--text-muted);margin-bottom:4px;">${frappe.utils.escape_html(label)}</div>
          <div style="font-size:26px;font-weight:700;color:${color};">${frappe.utils.escape_html(String(value))}</div>
          ${sub ? `<div style="font-size:11px;color:var(--text-muted);">${frappe.utils.escape_html(sub)}</div>` : ""}
        </div>
      `);
    });
  }

  // ── Stage Velocity (horizontal bar) ───────────────────────────────────
  _renderVelocity(rows) {
    const $el = this.page.main.find(".sd-velocity").empty();
    if (!rows.length) {
      $el.html(`<div style="color:var(--text-muted);padding:8px">${__("No ledger data for this period")}</div>`);
      return;
    }

    const MAX_AVG = Math.max(...rows.map(r => r.avg_days)) || 1;
    const TOP_N = rows.slice(0, 12); // max 12 stages

    let html = `<table style="width:100%;font-size:12px;border-collapse:collapse;">`;
    TOP_N.forEach(r => {
      const pct = Math.round((r.avg_days / MAX_AVG) * 100);
      const color = r.avg_days > 14 ? "#ef4444" : r.avg_days > 7 ? "#f59e0b" : "#10b981";
      html += `
        <tr style="border-bottom:1px solid var(--border-color);">
          <td style="width:130px;padding:5px 8px 5px 0;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;max-width:130px;"
              title="${frappe.utils.escape_html(r.state)}">${frappe.utils.escape_html(r.state)}</td>
          <td style="padding:5px 4px;">
            <div style="background:var(--bg-color);border-radius:4px;height:14px;width:100%;position:relative;">
              <div style="background:${color};border-radius:4px;height:14px;width:${pct}%;"></div>
            </div>
          </td>
          <td style="width:90px;text-align:right;padding:5px 0 5px 6px;font-weight:600;color:${color};">
            ${r.avg_days}d avg
          </td>
          <td style="width:60px;text-align:right;padding:5px 0 5px 6px;color:var(--text-muted);">
            (${r.transitions})
          </td>
        </tr>`;
    });
    html += `</table>`;
    $el.html(html);
  }

  // ── Pipeline Snapshot (state counts) ──────────────────────────────────
  _renderPipeline(data) {
    const $el = this.page.main.find(".sd-pipeline").empty();
    const rows = data.states || [];
    if (!rows.length) {
      $el.html(`<div style="color:var(--text-muted);padding:8px">${__("No active leads")}</div>`);
      return;
    }

    const STATE_COLORS = {
      "Signed": "#10b981", "Installed": "#22c55e", "Converted": "#0ea5e9",
      "Approved": "#f59e0b", "Rejected": "#ef4444", "Draft": "#6b7280",
      "Pending": "#3b82f6", "Agreement Sent": "#005c08", "Cancelled": "#9ca3af",
    };
    const MAX = Math.max(...rows.map(r => r.count)) || 1;
    const total = data.total || 1;

    let html = `<table style="width:100%;font-size:12px;border-collapse:collapse;">`;
    rows.slice(0, 15).forEach(r => {
      const pct = Math.round((r.count / MAX) * 100);
      const share = Math.round((r.count / total) * 100);
      const color = STATE_COLORS[r.state] || "#6366f1";
      html += `
        <tr style="border-bottom:1px solid var(--border-color);">
          <td style="width:160px;padding:5px 8px 5px 0;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;max-width:160px;"
              title="${frappe.utils.escape_html(r.state)}">${frappe.utils.escape_html(r.state)}</td>
          <td style="padding:5px 4px;">
            <div style="background:var(--bg-color);border-radius:4px;height:12px;width:100%;position:relative;">
              <div style="background:${color};border-radius:4px;height:12px;width:${pct}%;"></div>
            </div>
          </td>
          <td style="width:50px;text-align:right;padding:5px 0 5px 6px;font-weight:600;">${r.count}</td>
          <td style="width:40px;text-align:right;padding:5px 0 5px 4px;color:var(--text-muted);">${share}%</td>
        </tr>`;
    });
    html += `</table>`;
    $el.html(html);
  }

  // ── Trend chart (monthly) ──────────────────────────────────────────────
  _renderTrend(data) {
    const $el = this.page.main.find(".sd-trend").empty();
    const cats = data.categories || [];
    const series = data.series || [];

    if (!cats.length) {
      $el.html(`<div style="color:var(--text-muted);padding:8px">${__("No trend data")}</div>`);
      return;
    }

    // Native chart using frappe Charts SVG (no ECharts dependency on this page)
    const canvas = document.createElement("canvas");
    canvas.width = $el.width() || 600;
    canvas.height = 240;
    $el.append(canvas);

    // Build simple HTML table chart as fallback (reliable without extra deps)
    $el.empty();
    const colW = Math.max(60, Math.floor(($el.width() || 600) / cats.length));

    // Find max for scaling
    const allVals = series.flatMap(s => s.data);
    const MAX = Math.max(...allVals, 1);

    let html = `<div style="overflow-x:auto;"><div style="display:flex;align-items:flex-end;gap:2px;height:180px;padding:0 4px;">`;
    cats.forEach((cat, ci) => {
      html += `<div style="display:flex;flex-direction:column;align-items:center;flex:1;min-width:40px;">
        <div style="display:flex;align-items:flex-end;gap:2px;height:160px;">`;
      series.forEach(s => {
        const val = s.data[ci] || 0;
        const h = Math.round((val / MAX) * 140);
        html += `<div title="${frappe.utils.escape_html(s.name)}: ${val}"
                      style="width:10px;height:${h}px;background:${s.color || "#6b7280"};border-radius:2px 2px 0 0;"></div>`;
      });
      html += `</div><div style="font-size:10px;color:var(--text-muted);margin-top:4px;transform:rotate(-45deg);white-space:nowrap;">${frappe.utils.escape_html(cat)}</div></div>`;
    });
    html += `</div>`;

    // Legend
    html += `<div style="display:flex;gap:12px;margin-top:24px;flex-wrap:wrap;padding:0 4px;">`;
    series.forEach(s => {
      html += `<div style="display:flex;align-items:center;gap:4px;font-size:11px;">
        <div style="width:12px;height:12px;background:${s.color || "#6b7280"};border-radius:2px;"></div>
        ${frappe.utils.escape_html(s.name)}
      </div>`;
    });
    html += `</div></div>`;
    $el.html(html);
  }

  // ── Agent Attribution table ────────────────────────────────────────────
  _renderAttribution(rows) {
    const $el = this.page.main.find(".sd-attribution").empty();
    if (!rows.length) {
      $el.html(`<div style="color:var(--text-muted);padding:8px">${__("No signed records in this period")}</div>`);
      return;
    }

    let html = `
      <div style="overflow:auto;max-height:320px;">
      <table class="table table-bordered" style="margin:0;font-size:12px;">
        <thead style="position:sticky;top:0;background:#f9fafb;">
          <tr>
            <th>${__("Agent")}</th>
            <th title="${__("Lead was theirs at sign time")}">${__("Lead Agent")}</th>
            <th title="${__("Manager assigned them commission")}">${__("Closer")}</th>
            <th title="${__("Closer - Lead Agent")}">Δ</th>
          </tr>
        </thead><tbody>`;

    rows.forEach(r => {
      const diff = r.diff || 0;
      const diffColor = diff > 0 ? "#10b981" : diff < 0 ? "#ef4444" : "var(--text-muted)";
      const diffStr = diff > 0 ? `+${diff}` : String(diff);
      html += `<tr>
        <td style="font-weight:500;">${frappe.utils.escape_html(r.display_name || r.agent)}</td>
        <td style="text-align:center;">${r.as_lead_agent}</td>
        <td style="text-align:center;font-weight:600;">${r.as_closing_agent}</td>
        <td style="text-align:center;color:${diffColor};font-weight:600;">${diffStr}</td>
      </tr>`;
    });

    html += `</tbody></table></div>
      <div style="font-size:11px;color:var(--text-muted);margin-top:6px;">
        ${__("Δ = Closer count minus Lead Agent count. Positive means agent closes more deals than they sourced.")}
      </div>`;
    $el.html(html);
  }

  // ── Pending Attribution table ──────────────────────────────────────────
  _renderPending(rows) {
    const $el = this.page.main.find(".sd-pending").empty();
    this.page.main.find(".sd-pending-badge").text(rows.length);

    if (!rows.length) {
      $el.html(`<div style="color:#10b981;padding:8px;font-weight:600;">
        ${frappe.utils.icon("check", "sm")} ${__("All Signs have a Closing Agent assigned. Nothing pending.")}
      </div>`);
      return;
    }

    let html = `<div style="overflow:auto;max-height:320px;">
      <table class="table table-bordered" style="margin:0;font-size:12px;">
        <thead style="position:sticky;top:0;background:#fff0f0;">
          <tr>
            <th>${__("Signs Record")}</th>
            <th>${__("Lead")}</th>
            <th>${__("Business")}</th>
            <th>${__("Lead Agent")}</th>
            <th>${__("Sign Date")}</th>
            <th>${__("Action")}</th>
          </tr>
        </thead><tbody>`;

    rows.forEach(r => {
      html += `<tr>
        <td><a href="/app/signs/${encodeURIComponent(r.name)}" target="_blank">${frappe.utils.escape_html(r.name)}</a></td>
        <td><a href="/app/atm-leads/${encodeURIComponent(r.atm_leads)}" target="_blank">${frappe.utils.escape_html(r.atm_leads || "")}</a></td>
        <td>${frappe.utils.escape_html(r.business_name || "")}</td>
        <td>${frappe.utils.escape_html(r.lead_agent || "—")}</td>
        <td>${frappe.utils.escape_html(r.sign_date || "")}</td>
        <td>
          <button class="btn btn-xs btn-warning sd-assign-btn"
                  data-signs="${frappe.utils.escape_html(r.name)}"
                  style="font-size:11px;">
            ${__("Assign Agent")}
          </button>
        </td>
      </tr>`;
    });

    html += `</tbody></table></div>`;
    $el.html(html);

    // Quick-assign click handler: open Signs record in dialog
    $el.find(".sd-assign-btn").on("click", (e) => {
      const name = $(e.currentTarget).data("signs");
      this._quickAssign(name);
    });
  }

  // ── Quick-assign closing agent dialog ─────────────────────────────────
  _quickAssign(signsName) {
    const d = new frappe.ui.Dialog({
      title: __("Assign Closing Agent — {0}", [signsName]),
      fields: [
        {
          label: __("Closing Agent (Commission)"),
          fieldtype: "Link",
          options: "Sales Agent",
          fieldname: "closing_agent",
          reqd: 1,
        },
        {
          label: __("Commission Notes"),
          fieldtype: "Small Text",
          fieldname: "closing_notes",
        },
      ],
      primary_action_label: __("Save"),
      primary_action: (values) => {
        frappe.call({
          method: "frappe.client.set_value",
          args: {
            doctype: "Signs",
            name: signsName,
            fieldname: {
              closing_agent: values.closing_agent,
              closing_notes: values.closing_notes || "",
            },
          },
          callback: () => {
            frappe.show_alert({ message: __("Closing agent assigned"), indicator: "green" });
            d.hide();
            this.refresh();
          },
        });
      },
    });
    d.show();
  }
}

File: cclms/call_centre_lead_management_system/page/signs_dashboard/signs_dashboard.py
# Copyright (c) 2026, Galaxy and contributors
# For license information, please see license.txt
#
# Signs & Attribution Dashboard — backend APIs
# ─────────────────────────────────────────────
# Data sources:
#   • tabAgent Stage Ledger  — immutable transition log (days_in_prev_state, from/to_state)
#   • tabSigns               — sign records (lead agent snapshot + closing agent for commission)
#   • tabATM Leads           — live pipeline snapshot
#
# All APIs are @frappe.whitelist() and return plain dicts/lists.

import frappe
from frappe.utils import getdate, get_first_day, get_last_day, add_months


# ─────────────────────────────────────────────────────────────
# Helpers
# ─────────────────────────────────────────────────────────────

def _period(from_date=None, to_date=None):
    """Default to current month when no dates given."""
    today = getdate()
    start = getdate(from_date) if from_date else get_first_day(today)
    end   = getdate(to_date)   if to_date   else get_last_day(today)
    return str(start), str(end)


# ─────────────────────────────────────────────────────────────
# 1. Stage Velocity — avg days in each state before transitioning out
# ─────────────────────────────────────────────────────────────

@frappe.whitelist()
def get_stage_velocity(from_date=None, to_date=None, company=None, branch=None):
    """
    From Agent Stage Ledger: for each to_state, return
    avg days spent in the previous state (days_in_prev_state).

    This answers: "How long does it take us to move a lead through each stage?"
    """
    start, end = _period(from_date, to_date)

    where = ["asl.stage_date BETWEEN %(start)s AND %(end)s", "asl.days_in_prev_state > 0"]
    params = {"start": start, "end": end}

    if company:
        where.append("asl.company = %(company)s")
        params["company"] = company
    if branch:
        where.append("asl.branch = %(branch)s")
        params["branch"] = branch

    rows = frappe.db.sql(
        f"""
        SELECT
            asl.to_state                            AS state,
            ROUND(AVG(asl.days_in_prev_state), 1)   AS avg_days,
            COUNT(*)                                AS transitions,
            MIN(asl.days_in_prev_state)             AS min_days,
            MAX(asl.days_in_prev_state)             AS max_days
        FROM `tabAgent Stage Ledger` asl
        WHERE {' AND '.join(where)}
          AND asl.to_state NOT LIKE '[Sign Event]%%'
        GROUP BY asl.to_state
        ORDER BY avg_days DESC
        """,
        params,
        as_dict=True,
    )
    return rows


# ─────────────────────────────────────────────────────────────
# 2. Signs Summary — totals + pending attribution  
# ─────────────────────────────────────────────────────────────

@frappe.whitelist()
def get_signs_summary(from_date=None, to_date=None, company=None):
    """
    Returns:
    {
      total_signs         — Signs records created in period
      pending_attribution — Signs with no closing_agent (commission not assigned yet)
      attributed          — Signs with closing_agent filled
      pending_records     — [list] of pending Signs for the action table
    }
    """
    start, end = _period(from_date, to_date)

    where = ["sign_date BETWEEN %(start)s AND %(end)s"]
    params = {"start": start, "end": end}
    if company:
        where.append("company = %(company)s")
        params["company"] = company

    rows = frappe.db.sql(
        f"""
        SELECT
            name,
            atm_leads,
            sign_date,
            employee          AS lead_agent,
            closing_agent,
            company,
            branch,
            business_name,
            state_code
        FROM `tabSigns`
        WHERE {' AND '.join(where)}
        ORDER BY sign_date DESC
        """,
        params,
        as_dict=True,
    )

    pending = [r for r in rows if not r.get("closing_agent")]
    attributed = [r for r in rows if r.get("closing_agent")]

    return {
        "total_signs": len(rows),
        "pending_attribution": len(pending),
        "attributed": len(attributed),
        "pending_records": pending,
    }


# ─────────────────────────────────────────────────────────────
# 3. Agent Attribution — who created vs who closed
# ─────────────────────────────────────────────────────────────

@frappe.whitelist()
def get_agent_attribution(from_date=None, to_date=None, company=None):
    """
    Per Sales Agent shows:
    - as_lead_agent    : #Signs where they were the assigned agent at sign time
    - as_closing_agent : #Signs where manager selected them as the deal closer (commission)
    - diff             : closing - lead (positive = closer more deals than they brought)
    """
    start, end = _period(from_date, to_date)

    where = ["sign_date BETWEEN %(start)s AND %(end)s"]
    params = {"start": start, "end": end}
    if company:
        where.append("company = %(company)s")
        params["company"] = company

    cond = " AND ".join(where)

    lead_rows = frappe.db.sql(
        f"""
        SELECT employee AS agent, COUNT(*) AS cnt
        FROM `tabSigns`
        WHERE {cond} AND IFNULL(employee,'') != ''
        GROUP BY employee
        """,
        params, as_dict=True,
    )

    closer_rows = frappe.db.sql(
        f"""
        SELECT closing_agent AS agent, COUNT(*) AS cnt
        FROM `tabSigns`
        WHERE {cond} AND IFNULL(closing_agent,'') != ''
        GROUP BY closing_agent
        """,
        params, as_dict=True,
    )

    agent_map = {}

    def _ensure(a):
        if a not in agent_map:
            agent_map[a] = {"agent": a, "as_lead_agent": 0, "as_closing_agent": 0}
        return agent_map[a]

    for r in lead_rows:
        _ensure(r.agent)["as_lead_agent"] = r.cnt
    for r in closer_rows:
        _ensure(r.agent)["as_closing_agent"] = r.cnt

    # Resolve agent display names
    agent_ids = list(agent_map.keys())
    if agent_ids:
        sa_list = frappe.db.get_all(
            "Sales Agent",
            filters={"name": ["in", agent_ids]},
            fields=["name", "agent_name", "full_name"],
        )
        sa_map = {s.name: s for s in sa_list}
        for key, rec in agent_map.items():
            sa = sa_map.get(key)
            rec["display_name"] = (sa.full_name or sa.agent_name or key) if sa else key

    rows = list(agent_map.values())
    for r in rows:
        r["diff"] = r["as_closing_agent"] - r["as_lead_agent"]
        r.setdefault("display_name", r["agent"])

    rows.sort(key=lambda x: x["as_closing_agent"] + x["as_lead_agent"], reverse=True)
    return rows


# ─────────────────────────────────────────────────────────────
# 4. Pipeline Snapshot — live state counts from ATM Leads
# ─────────────────────────────────────────────────────────────

@frappe.whitelist()
def get_pipeline_snapshot(company=None, branch=None):
    """
    Live count of ATM Leads grouped by workflow_state.
    Also returns approval rate, rejection rate, conversion rate.
    """
    where = ["docstatus < 2"]
    params = {}
    if company:
        where.append("company = %(company)s")
        params["company"] = company
    if branch:
        where.append("branch = %(branch)s")
        params["branch"] = branch

    rows = frappe.db.sql(
        f"""
        SELECT
            IFNULL(workflow_state, 'Draft') AS state,
            COUNT(*) AS count
        FROM `tabATM Leads`
        WHERE {' AND '.join(where)}
        GROUP BY workflow_state
        ORDER BY count DESC
        """,
        params,
        as_dict=True,
    )

    state_map = {r.state: r.count for r in rows}
    total = sum(r.count for r in rows)

    def _pct(n):
        return round(n / total * 100, 1) if total else 0.0

    signed   = state_map.get("Signed", 0)
    approved = state_map.get("Approved", 0)
    rejected = state_map.get("Rejected", 0)
    installed = state_map.get("Installed", 0)

    return {
        "states": rows,
        "total": total,
        "sign_rate":     _pct(signed),
        "approval_rate": _pct(approved),
        "rejection_rate": _pct(rejected),
        "install_rate":  _pct(installed),
    }


# ─────────────────────────────────────────────────────────────
# 5. Approval vs Rejection trend — monthly from Agent Stage Ledger
# ─────────────────────────────────────────────────────────────

@frappe.whitelist()
def get_approval_rejection_trend(months_back=6, company=None):
    """
    Monthly counts of transitions TO: Approved, Rejected, Signed, Installed
    from Agent Stage Ledger — last N months.

    Returns: { categories: ["2026-01", ...], series: [{name, data}, ...] }
    """
    today = getdate()
    start_month = add_months(today.replace(day=1), -(int(months_back) - 1))

    where = ["stage_date >= %(start)s"]
    params = {"start": str(start_month)}
    if company:
        where.append("company = %(company)s")
        params["company"] = company

    rows = frappe.db.sql(
        f"""
        SELECT
            DATE_FORMAT(stage_date, '%%Y-%%m') AS month,
            to_state,
            COUNT(*) AS cnt
        FROM `tabAgent Stage Ledger`
        WHERE {' AND '.join(where)}
          AND to_state IN ('Approved','Rejected','Signed','Installed','Converted')
        GROUP BY month, to_state
        ORDER BY month ASC
        """,
        params,
        as_dict=True,
    )

    # Build month list
    categories = []
    cur = start_month
    while cur <= today:
        categories.append(f"{cur.year}-{cur.month:02d}")
        cur = add_months(cur, 1)

    series_map = {
        "Approved":  [0] * len(categories),
        "Rejected":  [0] * len(categories),
        "Signed":    [0] * len(categories),
        "Installed": [0] * len(categories),
        "Converted": [0] * len(categories),
    }
    cat_idx = {c: i for i, c in enumerate(categories)}

    for r in rows:
        if r.to_state in series_map and r.month in cat_idx:
            series_map[r.to_state][cat_idx[r.month]] += r.cnt

    COLORS = {
        "Approved":  "#f59e0b",
        "Rejected":  "#ef4444",
        "Signed":    "#10b981",
        "Installed": "#22c55e",
        "Converted": "#0ea5e9",
    }

    series = [
        {"name": k, "data": v, "color": COLORS.get(k, "#6b7280")}
        for k, v in series_map.items()
    ]

    return {"categories": categories, "series": series}
